File: cp1/podolianko_fb-11_cp1/src/main.rs
use std::env;
use std::fs::File;
use std::io::{BufReader, Read, Seek};

use podolianko_fb_11_cp1::*;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: pass text file location as the ony argument");
        std::process::exit(1);
    }

    let input_file = &args[1];

    println!("Current dir: {}",
             env::current_dir()?
                 .to_str()
                 .unwrap_or("Unknown"));
    println!("Sample file: {input_file}");

    println!();

    let mut with_whitespace = String::new();
    let mut without_whitespace = String::new();
    let mut reader = BufReader::new(File::open(input_file)?);
    reader.read_to_string(&mut with_whitespace)?;
    reader.rewind()?;
    reader.read_to_string(&mut without_whitespace)?;

    // read to String and error on invalid Unicode
    with_whitespace = normalize_string(&with_whitespace, VALID_LETTERS, true);
    without_whitespace = normalize_string(&without_whitespace, VALID_LETTERS, false);

    // Letter frequency calculations

    // With whitespaces
    let pmap = ngram_frequencies(&with_whitespace, 1, false)?;
    let entropy = entropy_from_probabilities(pmap.values());

    println!("Letter frequencies with whitespaces:\n");
    print_freq_plain(&pmap, 5, 4);
    println!("Entropy: {entropy}", entropy = entropy);
    println!("Redundancy: {}", redundancy(entropy, ((VALID_LETTERS.chars().count() as f64 + 1f64) as f64 ).log2()));
    println!();

    // Without whitespaces
    let pmap = ngram_frequencies(&without_whitespace, 1, false)?;
    let entropy = entropy_from_probabilities(pmap.values());

    println!("Letter frequencies without whitespaces:\n");
    print_freq_plain(&pmap, 5, 4);
    println!("Entropy: {entropy}", entropy = entropy);
    println!("Redundancy: {}", redundancy(entropy, (VALID_LETTERS.chars().count() as f64 ).log2()));
    println!();

    // Bigram frequency calculations

    // Overlapping, whitespaces
    let pmap = ngram_frequencies(&with_whitespace, 2, true)?;
    let entropy = entropy_from_probabilities(pmap.values());

    println!("Bigram frequencies overlapping, with whitespace:\n");
    print_bigram_freq_table(&pmap);
    let entropy_per_char = entropy / 2f64;
    println!("Entropy per character: {}", entropy_per_char);
    println!("Redundancy: {}", redundancy(entropy_per_char, ((VALID_LETTERS.chars().count() as f64 + 1f64) as f64 ).log2()));
    println!();

    // Overlapping, without whitespaces
    let pmap = ngram_frequencies(&without_whitespace, 2, true)?;
    let entropy = entropy_from_probabilities(pmap.values());

    println!("Bigram frequencies overlapping, without whitespace:\n");
    print_bigram_freq_table(&pmap);
    let entropy_per_char = entropy / 2f64;
    println!("Entropy per character: {}", entropy_per_char);
    println!("Redundancy: {}", redundancy(entropy_per_char, (VALID_LETTERS.chars().count() as f64 as f64 ).log2()));
    println!();

    // Non-overlapping, whitespaces
    let pmap = ngram_frequencies(&with_whitespace, 2, false)?;
    let entropy = entropy_from_probabilities(pmap.values());

    println!("Bigram frequencies non-overlapping, with whitespace:\n");
    print_bigram_freq_table(&pmap);
    let entropy_per_char = entropy / 2f64;
    println!("Entropy per character: {}", entropy_per_char);
    println!("Redundancy: {}", redundancy(entropy_per_char, ((VALID_LETTERS.chars().count() as f64 + 1f64) as f64 ).log2()));
    println!();

    // Non-overlapping, without whitespaces
    let pmap = ngram_frequencies(&without_whitespace, 2, false)?;
    let entropy = entropy_from_probabilities(pmap.values());

    println!("Bigram frequencies non-overlapping, without whitespace:\n");
    print_bigram_freq_table(&pmap);
    let entropy_per_char = entropy / 2f64;
    println!("Entropy per character: {}", entropy_per_char);
    println!("Redundancy: {}", redundancy(entropy_per_char, (VALID_LETTERS.chars().count() as f64 ).log2()));
    println!();

    //

    Ok(())
}

File: cp1/podolianko_fb-11_cp1/src/main.py
import sys
from lib import *


def main(file: str):
    with open(file, 'rt', encoding='utf-8') as textfile:
        sample_text_string = textfile.read()

    configs = [
        # W/wo whitespace , w/wo overlapping
        (True, True),
        (False, True),
        (True, False),
        (False, False)
    ]

    for whitespace_inclusion in (True, False):
        letter_executor(whitespace_inclusion, sample_text_string)

    for config in configs:
        bigram_executor(*config, sample_text_string)


def letter_executor(whitespaces: bool, text: str):
    text = normalize_string(text, VALID_LETTERS, whitespaces)
    pmap = ngram_frequencies(text, 1, False)
    entropy = entropy_from_probabilities(pmap.values())

    print(f"{'Letter frequencies with whitespaces:' if whitespaces else 'Letter frequencies without whitespaces'}")
    print()
    print_freq_plain(pmap, 5, 4)
    print(f"Entropy: {entropy}")
    print(f"Redundancy: {redundancy(entropy, log2(len(VALID_LETTERS) + (1 if whitespaces else 0)))}")
    print()


def bigram_executor(whitespaces: bool, overlapping: bool, text: str):
    text = normalize_string(text, VALID_LETTERS, whitespaces)
    pmap = ngram_frequencies(text, 2, overlapping)

    print(
        f"Bigram frequencies, {'overlapping' if overlapping else 'non-overlapping'}, {'with' if whitespaces else 'without'} whitespaces:")
    print()
    print_bigram_freq_table(pmap)
    entropy_per_char = entropy_from_probabilities(pmap.values()) / 2
    print(f"Entropy per character: {entropy_per_char}")
    print(f"Redundancy: {redundancy(entropy_per_char, log2(len(VALID_LETTERS) + (1 if whitespaces else 0)))}")
    print()


if __name__ == '__main__':
    if len(sys.argv) != 2:
        print("Usage: pass text file location as the only argument")
        exit(1)

    main(sys.argv[1])
